File: alchemist/alchemist-loading/src/main/java/it/unibo/alchemist/loader/export/MeanSquaredError.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.loader.export;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.apache.commons.math3.stat.descriptive.UnivariateStatistic;

import com.google.common.collect.Lists;

import it.unibo.alchemist.model.interfaces.Environment;
import it.unibo.alchemist.model.interfaces.Incarnation;
import it.unibo.alchemist.model.interfaces.Molecule;
import it.unibo.alchemist.model.interfaces.Reaction;
import it.unibo.alchemist.model.interfaces.Time;

/**
 * Exports the Mean Squared Error for the concentration of some molecule, given
 * another molecule that carries the correct result. The correct value is
 * extracted from every node, then the provided {@link UnivariateStatistic} is
 * applied to get a single, global correct value. Then, the actual value is
 * extracted from every node, its value is compared (subtracted) to the computed
 * correct value, it gets squared, and then logged.
 * 
 * @param <T>
 */
public final class MeanSquaredError<T> implements Extractor {

    private final Incarnation<T, ?> incarnation;
    private final String pReference;
    private final Molecule mReference;
    private final String pActual;
    private final Molecule mActual;
    private final List<String> name;
    private final UnivariateStatistic statistic;

    /**
     * @param localCorrectValueMolecule
     *            expected value {@link Molecule}
     * @param localCorrectValueProperty
     *            expected value property name
     * @param statistics
     *            the {@link UnivariateStatistic} to apply
     * @param localValueMolecule
     *            the target {@link Molecule}
     * @param localValueProperty
     *            the target property
     * @param incarnation
     *            the {@link Incarnation} to use
     */
    public MeanSquaredError(
            final Incarnation<T, ?> incarnation,
            final String localCorrectValueMolecule,
            final String localCorrectValueProperty,
            final String statistics,
            final String localValueMolecule,
            final String localValueProperty) {
        final Optional<UnivariateStatistic> statOpt = StatUtil.makeUnivariateStatistic(statistics);
        if (!statOpt.isPresent()) {
            throw new IllegalArgumentException("Could not create univariate statistic " + statistics);
        }
        statistic = statOpt.get();
        this.incarnation = incarnation;
        this.mReference = incarnation.createMolecule(localCorrectValueMolecule);
        this.pReference = localCorrectValueProperty == null ? "" : localCorrectValueProperty;
        this.pActual = localValueProperty == null ? "" : localValueProperty;
        this.mActual = incarnation.createMolecule(localValueMolecule);
        final StringBuilder mse = new StringBuilder("MSE(");
        mse.append(statistics);
        mse.append('(');
        if (!pReference.isEmpty()) {
            mse.append(pReference);
            mse.append('@');
        }
        mse.append(localCorrectValueMolecule);
        mse.append("),");
        if (!pActual.isEmpty()) {
            mse.append(pActual);
            mse.append('@');
        }
        mse.append(localValueMolecule);
        mse.append(')');
        name = Collections.unmodifiableList(Lists.newArrayList(mse.toString()));
    }

    @Override
    public double[] extractData(final Environment<?, ?> env, final Reaction<?> r, final Time time, final long step) {
        @SuppressWarnings("unchecked")
        final Environment<T, ?> environment = (Environment<T, ?>) env;
        final double value = statistic.evaluate(
                environment.getNodes().parallelStream()
                    .mapToDouble(n -> incarnation.getProperty(n, mReference, pReference))
                    .toArray());
        final double mse = environment.getNodes().parallelStream()
                .mapToDouble(n -> incarnation.getProperty(n, mActual, pActual) - value)
                .map(v -> v * v)
                .average()
                .orElseGet(() -> Double.NaN);
        return new double[]{mse};
    }

    @Override
    public List<String> getNames() {
        return name;
    }

}

File: alchemist-loading/src/main/kotlin/it/unibo/alchemist/boundary/extractors/MeanSquaredError.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.boundary.extractors

import it.unibo.alchemist.model.Actionable
import it.unibo.alchemist.model.Environment
import it.unibo.alchemist.model.Incarnation
import it.unibo.alchemist.model.Molecule
import it.unibo.alchemist.model.Time
import it.unibo.alchemist.util.StatUtil
import org.apache.commons.math3.stat.descriptive.UnivariateStatistic

/**
 * Exports the Mean Squared Error for the concentration of some molecule, given
 * another molecule that carries the correct result. The correct value is
 * extracted from every node, then the provided {@link UnivariateStatistic} is
 * applied to get a single, global correct value. Then, the actual value is
 * extracted from every node, its value is compared (subtracted) to the computed
 * correct value, it gets squared, and then logged.
 *
 * @param <T> concentration type
 */
class MeanSquaredError<T> @JvmOverloads constructor(
    incarnation: Incarnation<T, *>,
    localCorrectValueMolecule: String,
    localCorrectValueProperty: String = "",
    statistics: String,
    localValueMolecule: String,
    localValueProperty: String = "",
    precision: Int? = null,
) : AbstractDoubleExporter(precision) {

    constructor(
        incarnation: Incarnation<T, *>,
        localCorrectValueMolecule: String,
        statistics: String,
        localValueMolecule: String,
        precision: Int,
    ) : this(
        incarnation = incarnation,
        localCorrectValueMolecule = localCorrectValueMolecule,
        statistics = statistics,
        localValueMolecule = localValueMolecule,
        localValueProperty = "",
        precision = precision,
    )

    private val statistic: UnivariateStatistic = StatUtil.makeUnivariateStatistic(statistics)
        .orElseThrow { IllegalArgumentException("Could not create univariate statistic $statistics") }
    private val mReference: Molecule = incarnation.createMolecule(localCorrectValueMolecule)
    private val pReference: String = localCorrectValueProperty
    private val pActual: String = localValueProperty
    private val mActual: Molecule = incarnation.createMolecule(localValueMolecule)
    private val name: String = with(StringBuilder("MSE(")) {
        append(statistics)
        append('(')
        if (pReference.isNotEmpty()) {
            append(pReference).append('@')
        }
        append(localCorrectValueMolecule).append("),")
        if (pActual.isNotEmpty()) {
            append(pActual).append('@')
        }
        append(localValueMolecule).append(')')
        toString()
    }
    override val columnNames = listOf(name)

    override fun <T> extractData(
        environment: Environment<T, *>,
        reaction: Actionable<T>?,
        time: Time,
        step: Long,
    ): Map<String, Double> {
        val incarnation: Incarnation<T, *> = environment.incarnation
        val value: Double = statistic
            .evaluate(environment.nodes.map { incarnation.getProperty(it, mReference, pReference) }.toDoubleArray())
        val mse: Double = environment.nodes.parallelStream()
            .mapToDouble { incarnation.getProperty(it, mActual, pActual) - value }
            .map { it * it }
            .average()
            .orElse(Double.NaN)
        return mapOf(name to mse)
    }
}
